pkg/controller/master/virtualmachine/vmi_network_controller.go

```diff
@@ -12,107 +13,179 @@ import (
+		if iface.Name != "" && iface.MAC != "" {
 			vmiInterfaces[iface.Name] = iface.MAC
 		}
 	}
+	return vmiInterfaces
+}
 
-	for i, vmIface := range vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces {
-		macAddress, ok := vmiInterfaces[vmIface.Name]
-		// only set the network mac address when it has no existing value
-		if ok && vmIface.MacAddress == "" {
-			logrus.Debugf("set VM %s management network %s macAddress to %s", vm.Name, vmIface.Name, macAddress)
-			vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces[i].MacAddress = macAddress
-		}
+func (h *VMNetworkController) syncControllerRevision(vmi *kubevirtv1.VirtualMachineInstance, retry *bool) (*kubevirtv1.VirtualMachineInstance, error) {
+	vm, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return vmi, err
 	}
 
-	// note: this function is related to vm/vmi, should always run on vmi change
-	if err := h.regenerateControllerRevision(vmi, vm); err != nil {
-		return err
+	if vm == nil || vm.DeletionTimestamp != nil {
+		return vmi, nil
 	}
 
-	if equality.Semantic.DeepEqual(vmCopy.Spec.Template.Spec.Domain.Devices, vm.Spec.Template.Spec.Domain.Devices) {
-		return nil
+	// already synced
+	if vm.Generation == vm.Status.ObservedGeneration {
+		return vmi, nil
 	}
-	if _, err := h.vmClient.Update(vmCopy); err != nil {
-		return err
+
+	if err := h.regenerateControllerRevision(vmi, vm, retry); err != nil {
+		return vmi, err
 	}
 
-	return nil
+	return vmi, nil
 }
 
-func (h *VMNetworkController) regenerateControllerRevision(vmi *kubevirtv1.VirtualMachineInstance, vm *kubevirtv1.VirtualMachine) error {
+func (h *VMNetworkController) regenerateControllerRevision(vmi *kubevirtv1.VirtualMachineInstance, vm *kubevirtv1.VirtualMachine, retry *bool) error {
 	crObj, err := h.crCache.Get(vmi.Namespace, vmi.Status.VirtualMachineRevisionName)
 	if err != nil {
-		return fmt.Errorf("error fetch controller revision object for vmi %s-%s: %v", vmi.Name, vmi.Namespace, err)
+		// controllerRevision has a different name than vm name, essential to log it's name in error
+		return fmt.Errorf("failed to fetch controllerRevision %s/%s: %w", vmi.Namespace, vmi.Status.VirtualMachineRevisionName, err)
 	}
 
 	revisionSpec := &vmwatch.VirtualMachineRevisionData{}
 	if err = json.Unmarshal(crObj.Data.Raw, revisionSpec); err != nil {
-		return err
+		return fmt.Errorf("failed to unmarshal controllerRevision %s/%s: %w", crObj.Namespace, crObj.Name, err)
 	}
-	if !equality.Semantic.DeepEqual(revisionSpec.Spec.Template, vm.Spec.Template) && vm.Generation != vm.Status.ObservedGeneration {
-		patch, patchGenError := patchVMRevision(vm)
-		if patchGenError != nil {
-			return fmt.Errorf("error during patch generation for vmi %s-%s: %v", vmi.Name, vmi.Name, patchGenError)
+
+	revisionSpecCopy := revisionSpec
+	vmInterfaceMacs := getVMInterfaceMacs(vm)
+	vmiInterfaceMacs := getVmiInterfaceMacs(vmi)
+	updated := false
+	for i, rcIface := range revisionSpec.Spec.Template.Spec.Domain.Devices.Interfaces {
```

Comment:
Is it possible for the vm or vmi to have interfaces that aren't present in the controller revision?

Reply:
controllerrevision is a copy of VM spec when VM is started, controllerrevision is (re)created right then , and it can't be modifed unless deleted&created again. 

when VM has new interface added or interface changed, then they are different, and kubevirt will update VM.generation to new, but observedgeneration is kept old (or vice versa). that's how we get a hint from the UI that there are generation differences.